internal/character/march7th/ult.go

```diff
@@ -0,0 +1,99 @@
+package march7th
+
+import (
+	"github.com/simimpact/srsim/internal/global/common"
+	"github.com/simimpact/srsim/pkg/engine/info"
+	"github.com/simimpact/srsim/pkg/engine/modifier"
+	"github.com/simimpact/srsim/pkg/key"
+	"github.com/simimpact/srsim/pkg/model"
+)
+
+const (
+	MarchUltFreeze = "march7th-ultfreeze"
+	Ult            = "march7th-ult"
+)
+
+func init() {
+	modifier.Register(MarchUltFreeze, modifier.Config{
+		Duration: 1,
+	})
+}
+
+var ultHits = []float64{0.25, 0.25, 0.25, 0.25}
+
+func (c *char) Ult(target key.TargetID, state info.ActionState) {
+	for hitIndex, hitRatio := range ultHits {
+		c.ultHit(hitIndex, hitRatio)
+	}
+
+	c.engine.EndAttack()
+
+	freezeChance := 0.5
+	if c.info.Traces["103"] {
+		freezeChance += 0.15
+	}
+
+	for _, freezeTarget := range c.engine.Enemies() {
+		successfullyApplied, _ := c.engine.AddModifier(freezeTarget, info.Modifier{
+			Name:   common.Freeze,
+			Source: c.id,
+			State: common.FreezeState{
+				DamagePercentage: ultFreeze[c.info.UltLevelIndex()],
+				DamageValue:      0,
+			},
+			Chance:   freezeChance,
+			Duration: 1,
+		})
+
+		if c.info.Eidolon >= 1 {
+			if successfullyApplied {
+				c.engine.ModifyEnergy(info.ModifyAttribute{
+					Key:    Ult,
+					Target: c.id,
+					Source: c.id,
+					Amount: 6,
+				})
+			}
+		}
```

Comment:
Want to move `ModifyEnergy` out of for loop to only be called once like datamine in the event that some other mechanic subscribes to energy change events and is based on event count.